test/cpp/util/channelz_sampler_test.cc

```diff
@@ -0,0 +1,181 @@
+#include "src/cpp/server/channelz/channelz_service.h"
+#include "src/proto/grpc/testing/test.grpc.pb.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/util/subprocess.h"
+#include "test/cpp/util/test_credentials_provider.h"
+
+using grpc::Channel;
+using grpc::ClientContext;
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::Status;
+
+#include <gtest/gtest.h>
+
+// Test variables
+std::string server_address("0.0.0.0:10000");
+std::string custom_credentials_type("INSECURE_CREDENTIALS");
+std::string sampling_times = "2";
+std::string sampling_interval_seconds = "5";
+std::string output_json("./output.json");
+
+// Creata an echo server - randomly delay 0.1 to 0.2 s
+class EchoServerImpl final : public grpc::testing::TestService::Service {
+  Status EmptyCall(::grpc::ServerContext* context,
+                   const grpc::testing::Empty* request,
+                   grpc::testing::Empty* response) {
+    srand(unsigned(time(0)));
+    unsigned int server_delay_microseconds = 100000;
+    server_delay_microseconds += rand() % server_delay_microseconds;
+    usleep(server_delay_microseconds);
+    return Status::OK;
+  }
+};
+
+// Run server in a thread
+void RunServer(gpr_event* done_ev) {
+  // register channelz service
+  ::grpc::channelz::experimental::InitChannelzService();
+
+  EchoServerImpl service;
+  grpc::EnableDefaultHealthCheckService(true);
+  grpc_impl::reflection::InitProtoReflectionServerBuilderPlugin();
+  grpc::ServerBuilder builder;
+  auto server_creds =
+      grpc::testing::GetCredentialsProvider()->GetServerCredentials(
+          custom_credentials_type);
+  builder.AddListeningPort(server_address, server_creds);
+
+  // forces channelz and channel tracing to be enabled.
+  builder.AddChannelArgument(GRPC_ARG_ENABLE_CHANNELZ, 1);
+  builder.AddChannelArgument(GRPC_ARG_MAX_CHANNEL_TRACE_EVENT_MEMORY_PER_NODE,
+                             1024);
+  builder.RegisterService(&service);
+  std::unique_ptr<Server> server(builder.BuildAndStart());
+  gpr_log(GPR_INFO, "Server listening on %s", server_address.c_str());
+  while (true) {
+    if (gpr_event_get(done_ev)) {
+      return;
+    }
+  }
+}
+
+// Run client in a thread - - set timeout as 0.15s
+void RunClient(std::string client_id, gpr_event* done_ev) {
+  grpc::ChannelArguments channel_args;
+  std::shared_ptr<grpc::ChannelCredentials> channel_creds =
+      grpc::testing::GetCredentialsProvider()->GetChannelCredentials(
+          custom_credentials_type, &channel_args);
+  std::unique_ptr<grpc::testing::TestService::Stub> stub =
+      grpc::testing::TestService::NewStub(
+          grpc::CreateChannel(server_address, channel_creds));
+  unsigned int client_echo_sleep_second = 1;
+
+  gpr_log(GPR_INFO, "Client %s is echoing!", client_id.c_str());
+  while (true) {
+    if (gpr_event_get(done_ev)) {
+      return;
+    }
+    sleep(client_echo_sleep_second);
```

Comment:
nit: put this sleep after the RPC?

Otherwise, when the channelz sampler runs on this process, the client likely hasn't made the first RPC yet

Reply:
make sense. fixed.